src/Interpreters/Cache/Metadata.cpp

```diff
@@ -670,9 +673,86 @@ void CacheMetadata::downloadImpl(FileSegment & file_segment, std::optional<Memor
     LOG_TEST(log, "Downloaded file segment: {}", file_segment.getInfoForLog());
 }
 
-void CacheMetadata::cancelDownload()
+void CacheMetadata::startup()
+{
+    download_threads.reserve(download_threads_num);
+    for (size_t i = 0; i < download_threads_num; ++i)
+    {
+        download_threads.emplace_back(std::make_shared<DownloadThread>());
+        download_threads.back()->thread = std::make_unique<ThreadFromGlobalPool>([this, thread = download_threads.back()] { downloadThreadFunc(thread->stop_flag); });
+    }
+    cleanup_thread = std::make_unique<ThreadFromGlobalPool>(std::function{ [this]{ cleanupThreadFunc(); }});
+}
+
+void CacheMetadata::shutdown()
 {
     download_queue->cancel();
+    cleanup_queue->cancel();
+
+    for (auto & download_thread : download_threads)
+    {
+        if (download_thread->thread && download_thread->thread->joinable())
+            download_thread->thread->join();
+    }
+    if (cleanup_thread && cleanup_thread->joinable())
+        cleanup_thread->join();
+}
+
+bool CacheMetadata::isBackgroundDownloadEnabled()
+{
+    return download_threads_num;
+}
+
+bool CacheMetadata::setBackgroundDownloadThreads(size_t threads_num)
+{
+    if (threads_num == download_threads_num)
+        return false;
+
+    if (threads_num > download_threads_num)
+    {
+        SCOPE_EXIT({ download_threads_num = download_threads.size(); });
+
+        size_t add_threads = threads_num - download_threads_num;
+        for (size_t i = 0; i < add_threads; ++i)
+        {
+            download_threads.emplace_back(std::make_shared<DownloadThread>());
+            try
+            {
+                download_threads.back()->thread = std::make_unique<ThreadFromGlobalPool>(
+                    [this, thread = download_threads.back()] { downloadThreadFunc(thread->stop_flag); });
+            }
+            catch (...)
+            {
+                download_threads.pop_back();
+                throw;
+            }
+        }
+    }
+    else if (threads_num < download_threads_num)
+    {
+        size_t remove_threads = download_threads_num - threads_num;
+
+        {
+            std::lock_guard lock(download_queue->mutex);
+            for (size_t i = 0; i < remove_threads; ++i)
+                download_threads[download_threads.size() - 1 - i]->stop_flag = true;
```

Comment:
Which threads are more likely to be in the running state? This removal method stops those added the most recently. I'm unsure if this matters.

Reply:
All the threads are equivalent. Each thread polls one element from mutual queue and then processes it.
See 
https://github.com/ClickHouse/ClickHouse/blob/12ed9952c4009a52183a88b50292e781db7f1641/src/Interpreters/Cache/Metadata.cpp#L516-L534
and when an element is added to a download queue, one worker is notified 
https://github.com/ClickHouse/ClickHouse/blob/12ed9952c4009a52183a88b50292e781db7f1641/src/Interpreters/Cache/Metadata.cpp#L468-L478